[FLINK-32073][checkpoint] Implement file merging in snapshot

AlexYinHan opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

As one part of FLIP-306, this PR enables file merging in snapshot by implementing the corresponding output&input streams for ```FileMergingSnapshotManager```


## Brief change log
 - *Add ```createCheckpointStateOutputStream``` in ```FileMergingSnapshotManager```.*
 - *Add output&input streams for file merging checkpoints*
 - *Add ```SegmentFileStateHandle``` to represent a state that is written into a file segment.*


## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change added several UT tests in:
 - ```FileMergingSnapshotManagerTest```
 - ```FileMergingCheckpointStateOutputStreamTest```
 - ```FsSegmentDataInputStreamTest```.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

fredia commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerTest.java:
```diff
@@ -249,6 +261,104 @@ void testSizeStatsInPhysicalFile() throws IOException {
+    }
+
+    @Test
+    public void testConcurrentWriting() throws Exception {
```
I noticed that `FileMergingCheckpointStateOutputStream` is not thread-safe, how does this test guarantee thread safe?

AlexYinHan replied:
It is guaranteed by ```FileMergingSnapshotManager```. Only after the previous "Stream" is closed, will it create another "Stream" reusing the same physical file.

fredia commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerTest.java:
```diff
@@ -18,19 +18,28 @@
 import org.apache.flink.runtime.state.filesystem.AbstractFsCheckpointStorageAccess;
+import org.apache.flink.runtime.state.filesystem.FileMergingCheckpointStateOutputStream;
 
+import org.junit.jupiter.api.Assertions;
```
It‘s better to use AssertJ instead of Assertions.

https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#:~:text=Testing-,%23,-Tooling

AlexYinHan replied:
Thanks for the reminder. I have replaced Assertions with AssertJ.

fredia reviewed (changes requested):
@AlexYinHan Thanks for the PR, I left some comments, PTAL.

AlexYinHan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FileMergingCheckpointStateOutputStream.java:
```diff
@@ -0,0 +1,255 @@
+            // that the stream is closed
+            bufferPos = writeBuffer.length;
+            try {
+                fileMergingSnapshotManagerProxy.closeStreamExceptionally();
```
Answer to the first question: Yes. 

As declared in ```CheckpointStateOutputStream```, the ```close()``` method is intended to be the "unsuccessful close". For a "successful close",  you should call ```closeAndGetHandle()```.

AlexYinHan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/SegmentFileStateHandle.java:
```diff
@@ -0,0 +1,122 @@
+        return stateSize;
+    }
+
+    public long getEndPos() {
```
Indeed. I've removed that function.

Zakelly commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsSegmentDataInputStream.java:
```diff
@@ -0,0 +1,136 @@
+    private long markPosition;
+
+    private long curPosInFile;
+
```
I'm wondering if we could let this class simply proxy operations to the `FSDataInputStreamWrapper` and omit these members?

AlexYinHan replied:
Sure. This can simplify the code for ```FsSegmentDataInputStream```.

Note that, not all implementations of ```FSDataInputStream``` supports mark/reset. Therefore ```FsSegmentDataInputStream#markSupported()``` does not necessary return 'true' now.

Zakelly reviewed (changes requested):
@AlexYinHan Thanks for this PR! I left some comments PTAL.

AlexYinHan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FileMergingCheckpointStateOutputStream.java:
```diff
@@ -0,0 +1,255 @@
+        synchronized (this) {
+            if (!closed) {
+                try {
+                    flushToFile();
```
This class behaves similarly to ```FsCheckpointStateOutputStream```, which never calls ```outputStream.flush()```, but relies on ```outputStream.close()``` to ensure the persistency of data.

So, for ```flush()``` and ```sync()```, maybe we do not need to call ```outputStream.close()```.

Whereas we do need to think about ```closeAndGetHandle()```. I think we should call ```outputStream.sync()```  before returning the ```SegmentFileStateHandle``` instance, to make sure the file segment is visible. WDYT?

Zakelly replied:
If so, we'd better note this down since it violates the behavior of the base interface.

And for calling ```outputStream.sync()``` in ```closeAndGetHandle()```, I think it has been already covered by ```returnPhysicalFileForNextReuse```, WDYT?

AlexYinHan replied:
As for `closeAndGetHandle()`: Yeah, you are right. No need for an extra `outputStream.sync()` as the `returnPhysicalFileForNextReuse ` can handle it. I'd forgotten that.

AlexYinHan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerBase.java:
```diff
@@ -206,6 +213,64 @@ protected PhysicalFile createPhysicalFile(SubtaskKey subtaskKey, CheckpointedSta
     }
 
+    @Override
+    public FileMergingCheckpointStateOutputStream createCheckpointStateOutputStream(
```
It should be created by a ``` FsCheckpointStreamFactory```, similar to ```FsCheckpointStateOutputStream```. 

I have added a class named ```FsMergingCheckpointStorageLocation```, whose object is created by ```FsMergingCheckpointStorageAccess``` and creates ```FileMergingCheckpointStateOutputStream```.

AlexYinHan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FileStateHandle.java:
```diff
@@ -42,7 +42,7 @@ public class FileStateHandle implements StreamStateHandle {
 
     /** The size of the state in the file. */
-    private final long stateSize;
+    protected final long stateSize;
```
Make sense. ```SegmentFileStateHandle``` now directly implements ```StreamStateHandle```

AlexYinHan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FileMergingCheckpointStateOutputStream.java:
```diff
@@ -0,0 +1,255 @@
+                bufferPos = 0;
+            }
+        } else {
+            throw new IOException("closed");
```
Done.

AlexYinHan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FileMergingCheckpointStateOutputStream.java:
```diff
@@ -0,0 +1,255 @@
+    private long startPos;
+
+    /** current position relative to startPos. */
+    long curPosRelative = 0;
```
Yes. I have added the above information in the code comments.

AlexYinHan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FileMergingCheckpointStateOutputStream.java:
```diff
@@ -0,0 +1,255 @@
+     * A proxy of the {@link FileMergingSnapshotManager} that owns this {@link
+     * FileMergingCheckpointStateOutputStream}, with the interfaces for dealing with physical files.
+     */
+    public abstract static class FileMergingSnapshotManagerProxy {
```
Right. I've changed that into an interface.

AlexYinHan commented:
Thanks @fredia and @Zakelly for your comments. Sorry for the late reply. I have addressed your comments. Would you please take another look?

Zakelly commented:
> Thanks @fredia and @Zakelly for your comments. Sorry for the late reply. I have addressed your comments. Would you please take another look?

Thanks for your effort. I have no comments but a minor one, PTAL, thanks.

curcur commented:
CI seems Failed with "[99d7b97](https://github.com/apache/flink/pull/23514/commits/99d7b97cd5844ec6362b30d8f9134509b72cb647)"

AlexYinHan commented:
@flinkbot  run azure

AlexYinHan commented:
@curcur The CI failure has been addressed. (The file-merging checkpoint was unexpectedly enabled.) Would you please take another look?

curcur commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -517,24 +517,27 @@ protected StreamTask(
-                    e);
-            return checkpointStorageAccess;
-        }
+        // TODO (FLINK-32440): enable FileMergingCheckpoint by configuration
+        return checkpointStorageAccess;
+        //        if (fileMergingSnapshotManager == null) {
+        //            return checkpointStorageAccess;
+        //        }
+        //
+        //        try {
+        //            CheckpointStorageWorkerView mergingCheckpointStorageAccess =
+        //                    checkpointStorageAccess.toFileMergingStorage(
+        //                            fileMergingSnapshotManager, environment);
+        //            return (CheckpointStorageAccess) mergingCheckpointStorageAccess;
+        //        } catch (IOException e) {
+        //            LOG.warn(
+        //                    "Initiating FsMergingCheckpointStorageAccess failed"
+        //                            + "with exception: {}, falling back to original checkpoint
+        // storage access {}.",
+        //                    e.getMessage(),
+        //                    checkpointStorageAccess.getClass(),
+        //                    e);
+        //            return checkpointStorageAccess;
+        //        }
```
Can not leave code commented like this. Remove them if they are not needed.

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/EmptySegmentFileStateHandle.java:
```diff
@@ -0,0 +1,47 @@
+import java.io.UnsupportedEncodingException;
+
+/** An empty {@link SegmentFileStateHandle} that is only used as a placeholder. */
+public class EmptySegmentFileStateHandle extends SegmentFileStateHandle {
```
Is `EmptySegmentFileStateHandle` used in this PR?

If not, please remove it.

curcur commented:
The PR looks GTM.

1) I left a few comments.
2) squash the commits.
3) There are conflicts, resolve them and after turn green, ping me, and I will merge them.

AlexYinHan commented:
@curcur I have addressed your comments by removing the unused codes and squashing the commits. The PR is now free of conflicts and has passed the CI tests. PTAL.
